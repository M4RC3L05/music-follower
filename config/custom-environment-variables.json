{
  "apps": {
    "feed": {
      "host": "APPS_FEED_HOST",
      "port": {
        "__name": "APPS_FEED_PORT",
        "__format": "number"
      },
      "maxReleases": {
        "__name": "APPS_FEED_MAX_RELEASES",
        "__format": "number"
      }
    },
    "api": {
      "host": "APPS_API_HOST",
      "port": {
        "__name": "APPS_API_PORT",
        "__format": "number"
      },
      "basicAuth": {
        "username": "APPS_API_BASIC_AUTH_USERNAME",
        "password": "APPS_API_BASIC_AUTH_PASSWORD"
      }
    },
    "admin": {
      "host": "APPS_ADMIN_HOST",
      "port": {
        "__name": "APPS_ADMIN_PORT",
        "__format": "number"
      },
      "basicAuth": {
        "username": "APPS_ADMIN_BASIC_AUTH_USERNAME",
        "password": "APPS_ADMIN_BASIC_AUTH_PASSWORD"
      },
      "services": {
        "api": {
          "url": "APPS_ADMIN_SERVICES_API_URL",
          "basicAuth": {
            "username": "APPS_ADMIN_SERVICES_API_BASIC_AUTH_USERNAME",
            "password": "APPS_ADMIN_SERVICES_API_BASIC_AUTH_PASSWORD"
          }
        }
      }
    },
    "jobs": {
      "sync-releases": {
        "cron": {
          "pattern": "JOBS_SYNC_RELEASES_CRON_PATTERN",
          "timezone": "JOBS_SYNC_RELEASES_CRON_TIMEZONE",
          "tickerTimeout": {
            "__name": "JOBS_SYNC_RELEASES_CRON_TICKER_TIMEOUT",
            "__format": "number"
          }
        },
        "max-release-time": "JOBS_SYNC_RELEASES_MAX_RELEASE_TIME"
      }
    }
  },
  "database": {
    "path": "DATABASE_PATH"
  },
  "remote": {
    "itunes": {
      "search": {
        "url": "REMOTE_ITUNES_SEARCH_URL",
        "searchArtists": {
          "limit": {
            "__name": "REMOTE_ITUNES_SEARCH_SEARCH_ARTISTS_LIMIT",
            "__format": "number"
          }
        }
      },
      "lookup": {
        "url": "REMOTE_ITUNES_LOOKUP_URL",
        "getLatestsArtistMusicReleases": {
          "limit": {
            "__name": "REMOTE_ITUNES_LOOKUP_URL_GET_LATESTS_ARTIST_MUSIC_RELEASES_LIMIT",
            "__format": "number"
          }
        },
        "getLatestsArtistAlbumReleases": {
          "limit": {
            "__name": "REMOTE_ITUNES_LOOKUP_URL_GET_LATESTS_ARTIST_ALBUM_RELEASES_LIMIT",
            "__format": "number"
          }
        }
      }
    }
  },
  "media": {
    "placeholderImage": "MEDIA_PLACEHOLDER_IMAGE"
  }
}
