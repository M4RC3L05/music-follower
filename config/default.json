{
  "apps": {
    "rss": {
      "host": "0.0.0.0",
      "port": 4321
    },
    "admin": {
      "host": "0.0.0.0",
      "keys": ["foo", "bar"],
      "port": 4322,
      "session": {
        "maxAge": 300000,
        "httpOnly": true,
        "secure": false,
        "sameSite": "strict",
        "renew": true
      }
    },
    "sync-releases-task": {
      "cron": "0 */12 * * *"
    }
  },
  "database": {
    "client": "better-sqlite3",
    "useNullAsDefault": true,
    "connection": {
      "filename": "./database/db.sqlite"
    },
    "migrations": {
      "directory": "./database/migrations",
      "loadExtensions": [".mjs"],
      "extension": "mjs"
    }
  },
  "itunes": {
    "searchUrl": "https://itunes.apple.com/search",
    "lookup": "https://itunes.apple.com/lookup"
  },
  "media": {
    "placeholderImage": "https://via.placeholder.com/512"
  },
  "releases": {
    "albums": "https://itunes.apple.com/lookup?id={{id}}&entity=album&media=music&sort=recent&limit=70",
    "songs": "https://itunes.apple.com/lookup?id={{id}}&entity=song&media=music&sort=recent&limit=100"
  }
}
