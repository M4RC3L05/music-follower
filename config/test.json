{
  "apps": {
    "jobs": {
      "sync-releases": {
        "max-release-time": "1 year"
      }
    }
  },
  "database": {
    "path": ":memory:"
  },
  "logger": {
    "level": 4,
    "color": false
  }
}
