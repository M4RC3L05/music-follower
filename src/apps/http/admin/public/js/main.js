import { RouterProvider } from "react-router-dom";
import { StrictMode } from "react";
import { createRoot } from "react-dom/client";

import html from "./common/html.js";
import router from "./pages/mod.js";

createRoot(globalThis.document.querySelector("#app")).render(html`
  <${StrictMode}>
    <${RouterProvider} router=${router} />
  <//>
`);
