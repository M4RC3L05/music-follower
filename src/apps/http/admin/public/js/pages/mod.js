import { Route, createHashRouter, createRoutesFromElements, redirect } from "react-router-dom";

import Layout from "../components/layout.js";
import html from "../common/html.js";

const router = createHashRouter(
  createRoutesFromElements(html`
    <${Route} path="/" element=${html`<${Layout} />`}>
      <${Route} index lazy=${() => import("./index-page.js")} />
      <${Route} path="artists" lazy=${() => import("./artists-page.js")}>
        <${Route} path="remote" lazy=${() => import("./artists-remote-page.js")} />
        <${Route}
          path="reset"
          loader=${() => {
            redirect("/artists", { replace: true });
            return null;
          }}
        />
      <//>
      <${Route}
        path="releases"
        lazy=${() => import("./releases-page.js")}
        shouldRevalidate=${({ currentUrl, nextUrl }) =>
          !currentUrl.searchParams.has("selectedRelease") && !nextUrl.searchParams.has("selectedRelease")}
      />
    <//>
  `),
);

export default router;
