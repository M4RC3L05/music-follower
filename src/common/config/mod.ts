import vine from "@vinejs/vine";

const configSchema = vine.object({
  apps: vine.object({
    web: vine.object({
      host: vine.string(),
      port: vine.number(),
      basicAuth: vine.object({
        username: vine.string(),
        password: vine.string(),
      }),
      feed: vine.object({
        maxReleases: vine.number(),
      }),
    }),
    jobs: vine.object({
      syncReleases: vine.object({
        maxReleaseTime: vine.string(),
      }),
    }),
  }),
  database: vine.object({
    path: vine.string(),
  }),
  remote: vine.object({
    itunes: vine.object({
      search: vine.object({
        url: vine.string(),
        searchArtists: vine.object({
          limit: vine.number(),
        }),
      }),
      lookup: vine.object({
        url: vine.string(),
        getLatestsArtistMusicReleases: vine.object({
          limit: vine.number(),
        }),
        getLatestsArtistAlbumReleases: vine.object({
          limit: vine.number(),
        }),
      }),
    }),
  }),
  csrf: vine.object({
    origin: vine.string(),
  }),
  media: vine.object({
    remoteArtistsPlaceholderImage: vine.string(),
  }),
});
const configSchemaValidator = vine.compile(configSchema);

const config = await configSchemaValidator.validate({
  apps: {
    web: {
      host: Deno.env.get("APPS_WEB_HOST") ?? "127.0.0.1",
      port: Deno.env.get("APPS_WEB_PORT") ?? 4321,
      basicAuth: {
        username: Deno.env.get("APPS_WEB_BASIC_AUTH_USERNAME") ?? "foo",
        password: Deno.env.get("APPS_WEB_BASIC_AUTH_PASSWORD") ?? "bar",
      },
      feed: {
        maxReleases: Deno.env.get("APPS_WEB_FEED_MAX_RELEASES") ?? 50,
      },
    },
    jobs: {
      syncReleases: {
        maxReleaseTime: Deno.env.get("JOBS_SYNC_RELEASES_MAX_RELEASE_TIME") ??
          "1 year",
      },
    },
  },
  database: {
    path: Deno.env.get("DATABASE_PATH") ?? "./data/app.db",
  },
  remote: {
    itunes: {
      search: {
        url: Deno.env.get("REMOTE_ITUNES_SEARCH_URL") ??
          "https://itunes.apple.com/search",
        searchArtists: {
          limit: Deno.env.get("REMOTE_ITUNES_SEARCH_SEARCH_ARTISTS_LIMIT") ?? 8,
        },
      },
      lookup: {
        url: Deno.env.get("REMOTE_ITUNES_LOOKUP_URL") ??
          "https://itunes.apple.com/lookup",
        getLatestsArtistMusicReleases: {
          limit: Deno.env.get(
            "REMOTE_ITUNES_LOOKUP_URL_GET_LATESTS_ARTIST_MUSIC_RELEASES_LIMIT",
          ) ?? 70,
        },
        getLatestsArtistAlbumReleases: {
          limit: Deno.env.get(
            "REMOTE_ITUNES_LOOKUP_URL_GET_LATESTS_ARTIST_ALBUM_RELEASES_LIMIT",
          ) ?? 50,
        },
      },
    },
  },
  csrf: {
    origin: Deno.env.get("CSRF_ORIGIN") ?? "\\.*",
  },
  media: {
    remoteArtistsPlaceholderImage: Deno.env.get("MEDIA_PLACEHOLDER_IMAGE") ??
      "/public/images/remote-artist-image-default.svg",
  },
});

export default config;
