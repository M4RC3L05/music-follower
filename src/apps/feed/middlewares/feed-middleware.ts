import { sql } from "@m4rc3l05/sqlite-tag";
import config from "config";
import { Feed } from "feed";
import type { Context } from "@hono/hono";
import type { Release } from "#src/database/types/mod.ts";
import type {
  ItunesLookupAlbumModel,
  ItunesLookupSongModel,
} from "#src/remote/mod.ts";

export const feedMiddleware = (c: Context) => {
  const data = c.get("database").all<Release>(sql`
    select *
    from releases
    where "releasedAt" <= strftime('%Y-%m-%dT%H:%M:%fZ' , 'now')
      and not exists(
        select true from json_each(hidden)
        where json_each.value = 'feed'
      )
    order by "feedAt" desc
    limit ${config.get<number>("apps.feed.maxReleases")};
  `);

  const feed = new Feed({
    title: "Music releases",
    description: "Get the latest music releases from artist you follow",
    id: "music_follower",
    language: "en",
    copyright: "Music Follower",
    updated: new Date(),
    generator: "Music Follower",
  });

  for (const release of data) {
    const parsedMetadata = JSON.parse(release.metadata) as
      | ItunesLookupAlbumModel
      | ItunesLookupSongModel;

    feed.addItem({
      date: new Date(release.feedAt),
      description: `
      <img src="${release.coverUrl}" />
      <h1>${release.name}</h1><p>${release.artistName}</p>
      <p>${new Date(release.releasedAt).getFullYear()}</p>
      `.trim(),
      title: `${release.name} by ${release.artistName}`,
      id: String(release.id),
      link: release.type === "collection"
        ? parsedMetadata.collectionViewUrl ?? release.coverUrl
        : release.type === "track"
        ? (parsedMetadata as ItunesLookupSongModel).trackViewUrl ??
          parsedMetadata.collectionViewUrl ??
          release.coverUrl
        : release.coverUrl,
    });
  }

  const accepts = c.req.header("accept");

  if (
    accepts?.includes("application/rss+xml") ??
      accepts?.includes("application/xml")
  ) {
    return c.body(feed.rss2(), 200, {
      "content-type": accepts?.includes("application/xml")
        ? "application/xml"
        : "application/rss+xml",
    });
  }

  if (accepts?.includes("application/atom+xml")) {
    return c.body(feed.atom1(), 200, {
      "content-type": "application/atom+xml",
    });
  }

  if (accepts?.includes("application/json")) {
    return c.body(feed.json1(), 200, { "content-type": "application/json" });
  }

  return c.body(feed.rss2(), 200, { "content-type": "application/xml" });
};
