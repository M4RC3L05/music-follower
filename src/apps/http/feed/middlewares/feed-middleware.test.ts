/* eslint-disable @typescript-eslint/no-unsafe-argument */
import { beforeAll, beforeEach, describe, expect, it, vi } from "vitest";
import config from "config";

import * as fixtures from "#src/common/utils/test-utils/fixtures/mod.js";
import * as hooks from "#src/common/utils/test-utils/hooks/mod.js";
import { feedMiddleware } from "#src/apps/http/feed/middlewares/feed-middleware.js";

describe("feedMiddleware()", () => {
  beforeAll(async () => {
    await hooks.database.migrate();
  });

  beforeEach(() => {
    hooks.database.cleanup();

    vi.useFakeTimers({ now: 0 });
  });

  it("should fallback to rss if no accepts header is sent", async () => {
    const ctx = { request: { accepts: () => false } };

    feedMiddleware(ctx as any);

    expect((ctx as any).type).toBe("application/xml");
    expect((ctx as any).body).toMatchSnapshot();
  });

  it("should send rss feed if it accepts `application/xml`", async () => {
    const ctx = { request: { accepts: (x: any) => x === "application/xml" } };

    feedMiddleware(ctx as any);

    expect((ctx as any).type).toBe("application/xml");
    expect((ctx as any).body).toMatchSnapshot();
  });

  it("should send rss feed if it accepts `application/rss+xml`", async () => {
    const ctx = { request: { accepts: (x: any) => x === "application/rss+xml" } };

    feedMiddleware(ctx as any);

    expect((ctx as any).type).toBe("application/rss+xml");
    expect((ctx as any).body).toMatchSnapshot();
  });

  it("should send atom feed if it accepts `application/atom+xml`", async () => {
    const ctx = { request: { accepts: (x: any) => x === "application/atom+xml" } };

    feedMiddleware(ctx as any);

    expect((ctx as any).type).toBe("application/atom+xml");
    expect((ctx as any).body).toMatchSnapshot();
  });

  it("should send json feed if it accepts `application/json`", async () => {
    const ctx = { request: { accepts: (x: any) => x === "application/json" } };

    feedMiddleware(ctx as any);

    expect((ctx as any).type).toBe("application/json");
    expect((ctx as any).body).toMatchSnapshot();
  });

  for (const ct of ["application/rss+xml", "application/atom+xml", "application/json"]) {
    it(`should send feed with elements for "${ct}"`, () => {
      const ctx = { request: { accepts: (x: any) => x === ct } };

      fixtures.releases.load({ id: 1, feedAt: new Date(0) });
      fixtures.releases.load({ id: 2, feedAt: new Date(500) });
      fixtures.releases.load({ id: 3, feedAt: new Date(1000) });
      fixtures.releases.load({ id: 4, feedAt: new Date(1500) });

      vi.spyOn(config, "get").mockReturnValue(3);

      feedMiddleware(ctx as any);

      expect((ctx as any).type).toBe(ct);
      expect((ctx as any).body).toMatchSnapshot();
    });
  }

  it("should detect correctly the feed item link", async () => {
    const ctx = { request: { accepts: (x: any) => false } };

    fixtures.releases.load({
      id: 1,
      type: "track",
      metadata: { trackViewUrl: "foo", collectionViewUrl: "bar" },
      feedAt: new Date(0),
      coverUrl: "bix",
    });
    fixtures.releases.load({
      id: 2,
      type: "track",
      metadata: { trackViewUrl: undefined, collectionViewUrl: "bar" },
      feedAt: new Date(1),
      coverUrl: "bix",
    });
    fixtures.releases.load({
      id: 3,
      type: "track",
      metadata: { trackViewUrl: undefined, collectionViewUrl: undefined },
      feedAt: new Date(2),
      coverUrl: "bix",
    });
    fixtures.releases.load({
      id: 2,
      type: "collection",
      metadata: { collectionViewUrl: "foo" },
      feedAt: new Date(3),
      coverUrl: "bux",
    });
    fixtures.releases.load({
      id: 3,
      type: "collection",
      metadata: { collectionViewUrl: undefined },
      feedAt: new Date(4),
      coverUrl: "bux",
    });
    fixtures.releases.load({
      id: 1,
      type: "bix",
      metadata: { collectionViewUrl: undefined },
      feedAt: new Date(5),
      coverUrl: "bux",
    });

    vi.spyOn(config, "get").mockReturnValue(10);

    feedMiddleware(ctx as any);

    expect((ctx as any).type).toBe("application/xml");
    expect((ctx as any).body).toMatchSnapshot();
  });
});
