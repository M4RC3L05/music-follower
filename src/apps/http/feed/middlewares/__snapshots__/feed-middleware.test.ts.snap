// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`feedMiddleware() > should detect correctly the feed item link 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"utf-8\\"?>
<rss version=\\"2.0\\">
    <channel>
        <title>Music releases</title>
        <link>undefined</link>
        <description>Get the latest music releases from artist you follow</description>
        <lastBuildDate>Thu, 01 Jan 1970 00:00:00 GMT</lastBuildDate>
        <docs>https://validator.w3.org/feed/docs/rss2.html</docs>
        <generator>Music Follower</generator>
        <language>en</language>
        <copyright>Music Follower</copyright>
        <item>
            <title><![CDATA[bar by foo]]></title>
            <link>bux</link>
            <guid>1</guid>
            <pubDate>Thu, 01 Jan 1970 00:00:00 GMT</pubDate>
            <description><![CDATA[<img src=\\"bux\\" />
      <h1>bar</h1><p>foo</p>
      <p>1970</p>]]></description>
        </item>
        <item>
            <title><![CDATA[bar by foo]]></title>
            <link>bux</link>
            <guid>3</guid>
            <pubDate>Thu, 01 Jan 1970 00:00:00 GMT</pubDate>
            <description><![CDATA[<img src=\\"bux\\" />
      <h1>bar</h1><p>foo</p>
      <p>1970</p>]]></description>
        </item>
        <item>
            <title><![CDATA[bar by foo]]></title>
            <link>foo</link>
            <guid>2</guid>
            <pubDate>Thu, 01 Jan 1970 00:00:00 GMT</pubDate>
            <description><![CDATA[<img src=\\"bux\\" />
      <h1>bar</h1><p>foo</p>
      <p>1970</p>]]></description>
        </item>
        <item>
            <title><![CDATA[bar by foo]]></title>
            <link>bix</link>
            <guid>3</guid>
            <pubDate>Thu, 01 Jan 1970 00:00:00 GMT</pubDate>
            <description><![CDATA[<img src=\\"bix\\" />
      <h1>bar</h1><p>foo</p>
      <p>1970</p>]]></description>
        </item>
        <item>
            <title><![CDATA[bar by foo]]></title>
            <link>bar</link>
            <guid>2</guid>
            <pubDate>Thu, 01 Jan 1970 00:00:00 GMT</pubDate>
            <description><![CDATA[<img src=\\"bix\\" />
      <h1>bar</h1><p>foo</p>
      <p>1970</p>]]></description>
        </item>
        <item>
            <title><![CDATA[bar by foo]]></title>
            <link>foo</link>
            <guid>1</guid>
            <pubDate>Thu, 01 Jan 1970 00:00:00 GMT</pubDate>
            <description><![CDATA[<img src=\\"bix\\" />
      <h1>bar</h1><p>foo</p>
      <p>1970</p>]]></description>
        </item>
    </channel>
</rss>"
`;

exports[`feedMiddleware() > should fallback to rss if no accepts header is sent 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"utf-8\\"?>
<rss version=\\"2.0\\">
    <channel>
        <title>Music releases</title>
        <link>undefined</link>
        <description>Get the latest music releases from artist you follow</description>
        <lastBuildDate>Thu, 01 Jan 1970 00:00:00 GMT</lastBuildDate>
        <docs>https://validator.w3.org/feed/docs/rss2.html</docs>
        <generator>Music Follower</generator>
        <language>en</language>
        <copyright>Music Follower</copyright>
    </channel>
</rss>"
`;

exports[`feedMiddleware() > should send atom feed if it accepts \`application/atom+xml\` 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"utf-8\\"?>
<feed xmlns=\\"http://www.w3.org/2005/Atom\\">
    <id>music_follower</id>
    <title>Music releases</title>
    <updated>1970-01-01T00:00:00.000Z</updated>
    <generator>Music Follower</generator>
    <subtitle>Get the latest music releases from artist you follow</subtitle>
    <rights>Music Follower</rights>
</feed>"
`;

exports[`feedMiddleware() > should send feed with elements for "application/atom+xml" 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"utf-8\\"?>
<feed xmlns=\\"http://www.w3.org/2005/Atom\\">
    <id>music_follower</id>
    <title>Music releases</title>
    <updated>1970-01-01T00:00:00.000Z</updated>
    <generator>Music Follower</generator>
    <subtitle>Get the latest music releases from artist you follow</subtitle>
    <rights>Music Follower</rights>
    <entry>
        <title type=\\"html\\"><![CDATA[bar by foo]]></title>
        <id>4</id>
        <link href=\\"http://foo.bix\\"/>
        <updated>1970-01-01T00:00:01.500Z</updated>
        <summary type=\\"html\\"><![CDATA[<img src=\\"http://foo.bix\\" />
      <h1>bar</h1><p>foo</p>
      <p>1970</p>]]></summary>
    </entry>
    <entry>
        <title type=\\"html\\"><![CDATA[bar by foo]]></title>
        <id>3</id>
        <link href=\\"http://foo.bix\\"/>
        <updated>1970-01-01T00:00:01.000Z</updated>
        <summary type=\\"html\\"><![CDATA[<img src=\\"http://foo.bix\\" />
      <h1>bar</h1><p>foo</p>
      <p>1970</p>]]></summary>
    </entry>
    <entry>
        <title type=\\"html\\"><![CDATA[bar by foo]]></title>
        <id>2</id>
        <link href=\\"http://foo.bix\\"/>
        <updated>1970-01-01T00:00:00.500Z</updated>
        <summary type=\\"html\\"><![CDATA[<img src=\\"http://foo.bix\\" />
      <h1>bar</h1><p>foo</p>
      <p>1970</p>]]></summary>
    </entry>
</feed>"
`;

exports[`feedMiddleware() > should send feed with elements for "application/json" 1`] = `
"{
    \\"version\\": \\"https://jsonfeed.org/version/1\\",
    \\"title\\": \\"Music releases\\",
    \\"description\\": \\"Get the latest music releases from artist you follow\\",
    \\"items\\": [
        {
            \\"id\\": \\"4\\",
            \\"url\\": \\"http://foo.bix\\",
            \\"title\\": \\"bar by foo\\",
            \\"summary\\": \\"<img src=\\\\\\"http://foo.bix\\\\\\" />\\\\n      <h1>bar</h1><p>foo</p>\\\\n      <p>1970</p>\\",
            \\"date_modified\\": \\"1970-01-01T00:00:01.500Z\\"
        },
        {
            \\"id\\": \\"3\\",
            \\"url\\": \\"http://foo.bix\\",
            \\"title\\": \\"bar by foo\\",
            \\"summary\\": \\"<img src=\\\\\\"http://foo.bix\\\\\\" />\\\\n      <h1>bar</h1><p>foo</p>\\\\n      <p>1970</p>\\",
            \\"date_modified\\": \\"1970-01-01T00:00:01.000Z\\"
        },
        {
            \\"id\\": \\"2\\",
            \\"url\\": \\"http://foo.bix\\",
            \\"title\\": \\"bar by foo\\",
            \\"summary\\": \\"<img src=\\\\\\"http://foo.bix\\\\\\" />\\\\n      <h1>bar</h1><p>foo</p>\\\\n      <p>1970</p>\\",
            \\"date_modified\\": \\"1970-01-01T00:00:00.500Z\\"
        }
    ]
}"
`;

exports[`feedMiddleware() > should send feed with elements for "application/rss+xml" 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"utf-8\\"?>
<rss version=\\"2.0\\">
    <channel>
        <title>Music releases</title>
        <link>undefined</link>
        <description>Get the latest music releases from artist you follow</description>
        <lastBuildDate>Thu, 01 Jan 1970 00:00:00 GMT</lastBuildDate>
        <docs>https://validator.w3.org/feed/docs/rss2.html</docs>
        <generator>Music Follower</generator>
        <language>en</language>
        <copyright>Music Follower</copyright>
        <item>
            <title><![CDATA[bar by foo]]></title>
            <link>http://foo.bix</link>
            <guid>4</guid>
            <pubDate>Thu, 01 Jan 1970 00:00:01 GMT</pubDate>
            <description><![CDATA[<img src=\\"http://foo.bix\\" />
      <h1>bar</h1><p>foo</p>
      <p>1970</p>]]></description>
        </item>
        <item>
            <title><![CDATA[bar by foo]]></title>
            <link>http://foo.bix</link>
            <guid>3</guid>
            <pubDate>Thu, 01 Jan 1970 00:00:01 GMT</pubDate>
            <description><![CDATA[<img src=\\"http://foo.bix\\" />
      <h1>bar</h1><p>foo</p>
      <p>1970</p>]]></description>
        </item>
        <item>
            <title><![CDATA[bar by foo]]></title>
            <link>http://foo.bix</link>
            <guid>2</guid>
            <pubDate>Thu, 01 Jan 1970 00:00:00 GMT</pubDate>
            <description><![CDATA[<img src=\\"http://foo.bix\\" />
      <h1>bar</h1><p>foo</p>
      <p>1970</p>]]></description>
        </item>
    </channel>
</rss>"
`;

exports[`feedMiddleware() > should send json feed if it accepts \`application/json\` 1`] = `
"{
    \\"version\\": \\"https://jsonfeed.org/version/1\\",
    \\"title\\": \\"Music releases\\",
    \\"description\\": \\"Get the latest music releases from artist you follow\\",
    \\"items\\": []
}"
`;

exports[`feedMiddleware() > should send rss feed if it accepts \`application/rss+xml\` 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"utf-8\\"?>
<rss version=\\"2.0\\">
    <channel>
        <title>Music releases</title>
        <link>undefined</link>
        <description>Get the latest music releases from artist you follow</description>
        <lastBuildDate>Thu, 01 Jan 1970 00:00:00 GMT</lastBuildDate>
        <docs>https://validator.w3.org/feed/docs/rss2.html</docs>
        <generator>Music Follower</generator>
        <language>en</language>
        <copyright>Music Follower</copyright>
    </channel>
</rss>"
`;

exports[`feedMiddleware() > should send rss feed if it accepts \`application/xml\` 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"utf-8\\"?>
<rss version=\\"2.0\\">
    <channel>
        <title>Music releases</title>
        <link>undefined</link>
        <description>Get the latest music releases from artist you follow</description>
        <lastBuildDate>Thu, 01 Jan 1970 00:00:00 GMT</lastBuildDate>
        <docs>https://validator.w3.org/feed/docs/rss2.html</docs>
        <generator>Music Follower</generator>
        <language>en</language>
        <copyright>Music Follower</copyright>
    </channel>
</rss>"
`;
